Improve model builders

PepperCode1 opened the pull request:
- Additionally provide AO to material function
- Optimize mesh count by not preserving quad order between blocks
- Optimize mesh count by not splitting meshes when different inputs to the material function (except `RenderType`) produce the same material
- Add missing Nullable annotations

### TODO
- [x] Respect AO passed to new `ModelUtil.getMaterial` overload and add new `LightShader`s if necessary.
- [x] Optimize memory usage of `BufferBuilder`s.
- [x] Consider caching results of `BlockMaterialFunction` in `ModelBuilderResultConsumer` as `createKey` will be called a lot of times.

PepperCode1 committed:
Improve model builders

- Additionally provide AO to material function
- Optimize mesh count by not preserving quad order between blocks
- Optimize mesh count by not splitting meshes when different inputs to the material function (except RenderType) produce the same material

Jozufozu commented on common/src/lib/java/dev/engine_room/flywheel/lib/material/Materials.java:
```diff
@@ -10,43 +10,37 @@
 	public static final Material SOLID_BLOCK = SimpleMaterial.builder()
 			.build();
-	public static final Material SOLID_UNSHADED_BLOCK = SimpleMaterial.builder()
+	public static final Material SOLID_UNSHADED_BLOCK = SimpleMaterial.builderOf(SOLID_BLOCK)
```
Is adding new static final field in the middle binary compat safe?

PepperCode1 replied:
How a `static final` field is computed should never impact binary compat. Right now, no new fields are added, but adding new fields should not affect binary compat either as they are referenced by name.

Jozufozu commented on common/src/lib/java/dev/engine_room/flywheel/lib/model/baked/ResultConsumer.java:
```diff
@@ -0,0 +1,14 @@
+
+interface ResultConsumer<T> {
+	@Nullable
+	T createKey(RenderType renderType, boolean shade, boolean ambientOcclusion);
```
I feel like this doesn't need to be in this interface, or at least there should be a comment strongly justifying why it is.

Also since this is all internal, do we need the interface at all? It seems like there's only one concrete implementation and T is always Material. Do you anticipate needing this more complex behavior soon?

PepperCode1 replied:
The key mechanism allows combining geometry from different parameter values if they map to the same key. This function could be in a separate interface, but I think there's no point in extracting it since the other method in this interface accepts the key that `createKey` produces. Perhaps a better question is whether we need this mechanism at all. I do not know of any practical applications that would benefit from this, as the default `materialFunc` (`ModelUtil.getMaterial`) always has a unique mapping between input and output, but it could help with a non-standard material function that wants to ignore shade and/or AO.

Your second point is valid and something I considered while designing this. I originally designed `BakedModelBufferer` to be completely independent of other Flywheel classes, and I kept that constraint while making these refactors, but it should be reevaluated. I do not anticipate these classes being used for anything other than model builders now or in the future.

Jozufozu commented on forge/src/lib/java/dev/engine_room/flywheel/lib/model/baked/ForgeMeshEmitter.java:
```diff
@@ -0,0 +1,131 @@
+	}
+
+	/**
+	 * Some mods, like FramedBlocks, have custom hooks to determine the default AO. This method is invoked a second time
```
Good comment 🙂

Jozufozu commented on forge/src/lib/java/dev/engine_room/flywheel/lib/model/baked/ForgeMeshEmitterManager.java:
```diff
@@ -0,0 +1,13 @@
+
+import net.minecraft.client.renderer.RenderType;
+
+class ForgeMeshEmitterManager extends MeshEmitterManager<ForgeMeshEmitter> {
```
Is this subclass necessary?

PepperCode1 replied:
It encapsulates some logic, but probably not enough to justify a whole class. I'll inline it.

Jozufozu commented on common/src/lib/java/dev/engine_room/flywheel/lib/model/baked/MeshEmitterManager.java:
```diff
@@ -0,0 +1,78 @@
+class MeshEmitterManager<T extends MeshEmitter> {
+	private static final RenderType[] CHUNK_LAYERS = RenderType.chunkBufferLayers().toArray(RenderType[]::new);
+
+	final Reference2ReferenceMap<RenderType, T> emitterMap = new Reference2ReferenceLinkedOpenHashMap<>();
```
An ArrayMap may be better here

Jozufozu commented on common/src/lib/java/dev/engine_room/flywheel/lib/model/baked/MeshEmitterManager.java:
```diff
@@ -0,0 +1,78 @@
+
+	MeshEmitterManager(BiFunction<Supplier<BufferBuilder>, RenderType, T> meshEmitterFactory) {
+		for (RenderType renderType : CHUNK_LAYERS) {
+			T emitter = meshEmitterFactory.apply(this::getOrCreateBufferBuilder, renderType);
```
Each mesh emitter gets a separate function object here, but also I don't think the Supplier<BufferBuilder> needs to be a member variable of MeshEmitter.

You could store the buffer builder re-use logic in its own class, keep a final instance of it in this class, and pass it directly into MeshEmitter#getBuffer here: https://github.com/Engine-Room/Flywheel/pull/299/files#diff-1164194e4f2fd6cf77fb60e2b6cdbf10a46fbf79bbb990e3ecba404f29f3137dR61

Jozufozu commented on fabric/src/lib/java/dev/engine_room/flywheel/lib/model/baked/BakedModelBufferer.java:
```diff
@@ -110,46 +102,26 @@ public static void bufferBlocks(Iterator<BlockPos> posIterator, BlockAndTintGett
 				RenderType defaultLayer = ItemBlockRenderTypes.getChunkRenderType(state);
-				universalEmitter.prepare(defaultLayer);
-				model = universalEmitter.wrapModel(model);
+				boolean defaultAo = useAo && state.getLightEmission() == 0 && model.useAmbientOcclusion();
```
Is this logic copied from vanilla somewhere? Should we reference where so we can check for consistency later on?

Jozufozu commented on common/src/lib/java/dev/engine_room/flywheel/lib/model/baked/MeshEmitterManager.java:
```diff
@@ -0,0 +1,78 @@
+import it.unimi.dsi.fastutil.objects.ReferenceArrayList;
+import net.minecraft.client.renderer.RenderType;
+
+class MeshEmitterManager<T extends MeshEmitter> {
```
Very nice to have most of this consolidated in common now!

Jozufozu commented on common/src/lib/java/dev/engine_room/flywheel/lib/model/baked/BlockModelBuilder.java:
```diff
@@ -43,7 +43,17 @@ public BlockModelBuilder renderFluids(boolean renderFluids) {
+		return this;
+	}
+
+	public BlockModelBuilder materialFunc(@Nullable BlockMaterialFunction materialFunc) {
```
Binary compat?

Jozufozu committed:
self review

Jozufozu committed:
Direct result

- Completely inline result consumer
- MeshEmitterManager now directly creates the SimpleModel

Jozufozu committed:
Array of builders

- Inline ArrayList and directly use parallel arrays in MeshEmitter

Jozufozu committed:
Material AO

Jozufozu committed:
Move around and comment
